File: internal-base/mvps/src/main/java/com/lijing/dev/mvps/fragment/BaseAbstractFragment.java
package com.lijing.dev.mvps.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.kaopiz.kprogresshud.KProgressHUD;
import com.lijing.dev.mvps.presenter.BaseAbstractPresenter;

import javax.inject.Inject;

/**
 * 我们的目标是，尽量不用 fragment
 * 2.0 该用还是要用啊
 *
 * @author lijing
 */
public abstract class BaseAbstractFragment<P extends BaseAbstractPresenter> extends Fragment implements IBaseFragment {
    protected final String TAG = this.getClass().getSimpleName();

    @Inject
    protected P mPresenter;

    protected Activity mContext;

    protected View mRootView;

    protected KProgressHUD mProgressHUD;

//    protected Unbinder mUnbinder;

    @Override
    public void onAttach(Activity context) {
        super.onAttach(context);
        mContext = context;
        initInject();
        attachPresenter();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (getContentLayoutID() != 0) {
             mRootView = inflater.inflate(getContentLayoutID(), null);

            return mRootView;
        }
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
//        mUnbinder = ButterKnife.bind(this, mRootView);
        initVariables();
        initViews();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initEvents();
        initData();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mPresenter != null) {
            mPresenter.onSaveInstanceState(outState);
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        showHud(false);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
//        if (mUnbinder != null) {
//            mUnbinder.unbind();
//        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        detachPresenter();
    }

    /* presenter */

    private void attachPresenter() {
        if (mPresenter != null) {
            mPresenter.attachView(this);
        }
    }

    private void detachPresenter() {
        if (mPresenter != null) {
            mProgressHUD = null;
            mPresenter.unsubscribe();
            mPresenter.detachView();
        }
    }


    @Override
    public void showHud(boolean b) {
        if (mProgressHUD == null) {
            mProgressHUD = KProgressHUD
                    .create(mContext)
                    .setStyle(KProgressHUD.Style.SPIN_INDETERMINATE)
                    .setCancellable(true)
                    .setAnimationSpeed(2)
                    .setDimAmount(0.5f);
        }

        if (mProgressHUD.isShowing() && !b) {
            mProgressHUD.dismiss();
        } else if (!mProgressHUD.isShowing() && b) {
            mProgressHUD.show();
        }
    }

    @Override
    public void showMessage(String msg) {
        Toast.makeText(mContext, msg, Toast.LENGTH_SHORT).show();
    }

}

File: internal-base/lib-mvvm/src/main/java/com/lijing/dev/mvvm/core/BaseAbstractFragment.kt
package com.lijing.dev.mvvm.core

import android.content.Context
import android.os.Bundle
import android.os.Looper
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class BaseAbstractFragment<VM : BaseViewModel> : Fragment(), IBaseView<VM> {

    /**
     * 与 Activity 关联
     **/
    override fun onAttach(context: Context?) {
        super.onAttach(context)
        subscribeViewModel()
        attachViewModelInternal()
    }

    private fun attachViewModelInternal() {
        if (activity is BaseAbstractActivity<*>) {
            (activity as BaseAbstractActivity<*>).attachViewModelInternal(getViewModel())
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initVariables()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(getContentLayoutID(), null, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initViewsAndEvents()
        enqueueIdleTask()
    }

    override fun onStart() {
        super.onStart()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onStop() {
        super.onStop()
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onDetach() {
        super.onDetach()
    }

    override fun showHud(show: Boolean) {
        if (activity is BaseAbstractActivity<*>) {
            (activity as BaseAbstractActivity<*>).showHud(show)
        }
    }


    private fun enqueueIdleTask() {
        Looper.myQueue().addIdleHandler {
            idleTask()
            return@addIdleHandler false
        }
    }

    /**
     * 使用 idle handler 在主线程空闲时调用
     */
    open fun idleTask() {}

}